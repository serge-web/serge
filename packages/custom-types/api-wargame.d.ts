import { PlayerUiDispatch, PlayerUiActionTypes } from './player-ui-actions'
import Message from './message.d'

export type ApiWargameDb = PouchDB.Database<Message>

export interface ApiWargameListenNewMessage {
  db: ApiWargameDb,
  name: string,
  dispatch: PlayerUiDispatch
}

export type ApiWargameDbObject = { db: ApiWargameDb, name: string }
