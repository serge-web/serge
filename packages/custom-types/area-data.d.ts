import ForceData from "./force-data"


/** a user-defined area, used for organising activities */
export interface Area {
  // user name for this area
  name: string,
  // which forces use this area
  usedBy: Array<ForceData['uniqid']>
  // the bounds of the area
  polygon: GeoJSON.Polygon
}

/** the parent wrapper for the set of areas */
export default interface AreaData {
  areas: Array<Area>
}