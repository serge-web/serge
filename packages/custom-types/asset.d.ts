import { AttributeValues, Perception, PlatformTypeData, RouteStatus, RouteTurn } from '.'
import { LaydownTypes } from '@serge/config'
export interface Asset {
  /**
   * the name of this asset
   */
  name: string,
  /** when an asset is first identified, only the
   * contact id is shown.  Something like `C023`.
   */
  readonly contactId: string,
  /** the uniq id of this asset, present to allow
   * assets to be renamed, or multiple assets of same
   * name present in game
   */
  readonly uniqid: string,
  /** the current position of an asset. It's optional,
   * since an asset may be carried by another
   */
  position?: string,
  /** how other forces perceive this asset
   */
  perceptions: Array<Perception>,
  /** where this platform has already been
   */
  history?: Array<RouteTurn>,
  /** the set of planned turns for this asset
   */
  plannedTurns?: Array<RouteTurn>,
  /** the current condition of this asset
   */
  condition: string,
  /** 
   * the type-id of this platform 
   */
  platformTypeId: PlatformTypeData['uniqid']
  /** if the player is allowed to change the position of this asset
   * during turn-zero of a wargame
   * note: we allow string so it can be set in JSON
   */
  locationPending?: LaydownTypes | 'Force laydown' | 'Umpire laydown' 
  /** whether this asset has been destroyoed
   * //TODO: move away from this, use condition
   */
  destroyed?: boolean
  /** if this asset is actually an organisational group
   * of assets
   */
  comprising?: Array<Asset>
  /** if this asset is carrying other assets
   */
  hosting?: Array<Asset>
  /** current status of this asset
   *
  */
  status?: RouteStatus
  /** attributes for this asset 
   */
  attributeValues?: AttributeValues
}

export default Asset
