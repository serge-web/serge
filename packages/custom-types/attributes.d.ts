import {
  ATTRIBUTE_TYPE_ENUM,
  ATTRIBUTE_TYPE_NUMBER,
  ATTRIBUTE_TYPE_STRING,
  ATTRIBUTE_VALUE_ENUM,
  ATTRIBUTE_VALUE_NUMBER,
  ATTRIBUTE_VALUE_STRING
} from '@serge/config'

/** Types to support the storage of attributes within the game */

/** the basic type details common to all attributes */
export interface BaseAttributeType {
  attrType: string
  // human-readable name for this atribute type
  name: string
  // unique id for this atribute type
  attrId: string
  // description of this type of atribute
  description?: string
  // whether atribute can be edited by players
  editableByPlayer?: boolean
}

/** a number value */
export interface NumberAttributeType extends BaseAttributeType {
  attrType: typeof ATTRIBUTE_TYPE_NUMBER
  /** name of units to use for this type */
  units?: string
  /** default value to use for this atribute */
  defaultValue?: number
  /** how to display values for this atribute (.e.g `0.00`) */
  format?: string
}

/** a number value */
export interface StringAttributeType extends BaseAttributeType {
  attrType: typeof ATTRIBUTE_TYPE_STRING
  /** default value to use for this atribute */
  defaultValue?: string
  /** how to display values for this atribute (.e.g `0.00`) */
}

/** a string value from a list of possible values */
export interface EnumAttributeType extends BaseAttributeType {
  attrType: typeof ATTRIBUTE_TYPE_ENUM
  /** possible values for this attribute */
  values: string[]
  /** default value to use for this atribute */
  defaultValue?: string
}


/** the basic type details common to all attributes */
export interface BaseAttributeValue {
  attrType: string
  // id of this atribute (allows details to be looked up from type data)
  attrId: BaseAttributeType['attrId']
}

export interface StringAttributeValue extends BaseAttributeValue {
  attrType: typeof ATTRIBUTE_VALUE_STRING
  // value of this atribute
  value: string
}

export interface NumberAttributeValue extends BaseAttributeValue {
  attrType: typeof ATTRIBUTE_VALUE_NUMBER
  // value of this atribute
  value: number
}

export interface EnumAttributeValue extends BaseAttributeValue {
  attrType: typeof ATTRIBUTE_VALUE_ENUM
  // value of this atribute
  value: string
}

/** union of all types of atribute type */
export type AttributeType = NumberAttributeType | EnumAttributeType | StringAttributeType

/** union of all types of atribute value */
export type AttributeValue = NumberAttributeValue | EnumAttributeValue | StringAttributeValue

/** collection of instances of types of atribute */
export type AttributeTypes = AttributeType[]

/** collection of instances of values of atribute */
export type AttributeValues = AttributeValue[]

/** collate data for attribute editor table */
export interface AttributeEditorData {
  // name for read interface
  nameRead: string
  // name for text editor, optionally includes units
  nameWrite: string
  description?: string
  // value as displayed in read-only mode
  valueRead: string
  // value as supplied to text field
  valueWrite: string
  // attribute id, needed to store results
  attrId: string
  // value type, needed to store results
  valueType: typeof ATTRIBUTE_VALUE_NUMBER | typeof ATTRIBUTE_VALUE_ENUM | typeof ATTRIBUTE_VALUE_STRING
  // whether the player can edit the field
  playerCanEdit: boolean
}