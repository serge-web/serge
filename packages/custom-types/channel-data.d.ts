import Participant, { ParticipantCustom, ParticipantCollab, ParticipantChat, ParticipantTemplate, ParticipantMapping } from './participant'
import { SpecialChannelTypes, CHANNEL_CUSTOM, CHANNEL_COLLAB, CHANNEL_CHAT, CHANNEL_MAPPING, SpecialChannelColumns, InitialStates } from '@serge/config'

/** special configuration for collaborative channels */
export interface CollabOptions {  
  /** edit mode */
  // TODO: drop this, we use channel.format
  mode: 'edit' | 'response',
  /** verbs for returning for edit */
  returnVerbs: Array<string>,
  /** who original document goes to first */
  startWithReview : boolean,
  /** extra columns to show in status view */
  extraColumns: Array<SpecialChannelColumns>
}

/** description of channel, as stored in database */
// @deprecate to be deleted
export default interface ChannelData {
  /** unique id for this channel */
  readonly uniqid: string,
  /** name of this channel */
  name: string,
  /** list of participants for this channel */
  participants: Array<Participant>
  /** special channel types */
  format?: SpecialChannelTypes
  /** special collaborative working options */
  collabOptions?: CollabOptions
}

/** description of channel, as stored in database */
export interface ChannelCore {
  /** unique id for this channel */
  readonly uniqid: string
  /** name of this channel */
  name: string
}

/** description of chat channel */
export interface ChannelChat extends ChannelCore {
  channelType: typeof CHANNEL_CHAT
  /** list of participants for this channel */
  participants: Array<ParticipantChat>
}

/** description of mapping channel */
export interface ChannelMapping extends ChannelCore {
  channelType: typeof CHANNEL_MAPPING
  /** list of participants for this channel */
  participants: Array<ParticipantMapping>
}

/** description of standard channel, sending custom messages */
export interface ChannelCustom extends ChannelCore {
  channelType: typeof CHANNEL_CUSTOM
  /** list of participants for this channel */
  participants: Array<ParticipantCustom>
}

/** description of collaborative editing channel */
export interface ChannelCollab extends ChannelCore {
  channelType: typeof CHANNEL_COLLAB
  /** list of participants for this channel */
  participants: Array<ParticipantCollab>
  /** verbs for returning for edit */
  requestChangesVerbs: Array<string>
  /** verbs for approving document */
  approveVerbs: Array<string>
  /** verbs for releasing document */
  releaseVerbs: Array<string>
  /** who original document goes to first */
  initialState : InitialStates
  /** extra columns to show in status view */
  extraColumns: Array<SpecialChannelColumns>
  /** type for new document */
  newMessageTemplate: ParticipantTemplate
  /** type for response document */
  responseTemplate?: ParticipantTemplate
}

/** union of the assorted channel types */
export type ChannelTypes = ChannelChat |
  ChannelCustom |
  ChannelCollab | 
  ChannelMapping
