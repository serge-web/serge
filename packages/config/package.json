{
  "name": "@serge/config",
  "version": "0.3.0",
  "license": "Apache-2.0",
  "author": "foxleigh81 <alex@alexfoxleigh.com>",
  "main": "build/index.js",
  "dependencies": {
    "@types/dotenv": "^8.2.0",
    "@types/expired-storage": "^1.0.0",
    "dotenv": "^8.2.0",
    "expired-storage": "^1.0.2",
    "typescript": "^4.0.5"
  },
  "scripts": {
    "build": "tsc --pretty",
    "watch": "tsc --watch"
  }
}
