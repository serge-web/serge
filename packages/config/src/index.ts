/***
* This file is autogenerated, please do not edit it.
* To generate an updated version, please run `yarn genreg`
***/

export * from './globals'
export * from './routes'
export { Phase, LaydownPhases, PlanningStates, PlanningCommands } from './enums'
export * from './action-constants'
