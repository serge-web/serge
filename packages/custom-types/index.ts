/***
* This file is autogenerated, please do not edit it.
* To generate an updated version, please run `yarn genreg`
***/

export { default as Asset } from './asset.d'
export * from './attributes'
export * from './channel-data.d'
export { default as ChannelUI } from './channel-ui.d'
export { default as Channel } from './channel.d'
export { default as Editor } from './editor.d'
export * from './export-items-actions.d'
export * from './export-items.d'
export { default as ForceData } from './force-data.d'
export { default as ForceOption } from './force-option.d'
export { default as ForceRole } from './force-role.d'
export { default as Force } from './force.d'
export * from './forms.d'
export * from './global.d'
export { default as GroupItem } from './group-item.d'
export { default as IconOption } from './icon-option.d'
export * from './map-annotation'
export { default as MapPostBack } from './map-postback-data.d'
export { default as MappingConstraints } from './mapping-constraints.d'
export { default as MappingContext } from './mapping-context.d'
export { default as MarkerPerception } from './marker-perception.d'
export * from './message-types-actions.d'
export { default as MessageTypes, TemplateBodysByKey } from './message-types.d'
export {
  ChatMessage, CollaborationDetails, CoreMessage, default as Message, FeedbackItem, InteractionDetails
  // All template based messages
  , InteractionMessageStructure, MessageChannel, MessageCloneMarker, MessageCounter, MessageCreateTaskGroup,
  // All template based messages
  MessageCustom, MessageDeleteMarker, MessageDeletePlatform, MessageDetails,
  MessageDetailsFrom,
  // System Messages
  MessageFeedback,
  // Map Types
  MessageForceLaydown, MessageHostPlatform, MessageInfoType,
  MessageInfoTypeClipped, MessageInteraction, MessageLeaveTaskGroup, MessageMap, MessagePerceptionOfContact, MessagePlanning,
  MessageStateOfWorld, MessageStructure, MessageSubmitPlans, MessageUpdateMarker, MessageVisibilityChanges, PlanningMessageStructure
} from './message.d'
export { MessagesActionTypes, MessagesDispatch, SetCurrentViewFromURI } from './messages-action'
export * from './participant.d'
export { default as PerceivedTypes } from './perceived-types.d'
export { default as Perception } from './perception.d'
export { default as PlanMobileAsset, TurningDetails } from './plan-mobile-asset'
export { default as PlannedRoute } from './planned-route.d'
export * from './planning-activity'
export { default as PlatformTypeData } from './platform-type-data.d'
export { default as PlatformType } from './platform-type.d'
export * from './player-log-actions'
export * from './player-log.d'
export * from './player-ui-actions.d'
export { default as PlayerUi, PlayerUiChannels, PlayerUiChatChannel } from './player-ui.d'
export { default as RequestForInformation } from './request-for-information.d'
export { default as Role } from './role.d'
export * from './root-state'
export { default as RouteMarker, RouteMarkerDuo, RouteMarkerStatus as RouteMarkerStatus } from './route-marker'
export { default as RouteStatus } from './route-status.d'
export { default as RouteStore } from './route-store.d'
export { default as RouteTurn } from './route-turn'
export { default as Route, RouteChild } from './route.d'
export { default as SelectedAsset } from './selected-asset.d'
export * from './serge-hex-3.d'
export { default as SetWargameMessage } from './set-wargame-message.d'
export { AssetState, ForceState, StateOfWorld } from './state-of-world.d'
export { default as State } from './state.d'
export { default as Template, TemplateBody } from './template.d'
export * from './turn-length.d'
export { default as Visibility } from './visibility.d'
export { WargameActionTypes, WargameDispatch, WargameRevision } from './wargame-action'
export { default as WargameData } from './wargame-data.d'
export { default as WargameList } from './wargame-list.d'
export { default as WargameOverview } from './wargame-overview.d'
export { default as Wargame } from './wargame.d'

