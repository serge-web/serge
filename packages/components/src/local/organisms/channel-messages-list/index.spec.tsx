/* global it expect */
import React from 'react'
import renderer from 'react-test-renderer'
import ChannelMessagesList from './index'
import { CUSTOM_MESSAGE } from '@serge/config'
import { MessageCustom } from '@serge/custom-types'
import moment from 'moment-timezone'

describe('ChannelMessagesList component: ', () => {
  it('renders component correctly', () => {
    moment.tz.setDefault('Etc/UTC')
    const icons = [
      './images/default_img/forceDefault.png'
    ]
    const colors = [
      '#F00'
    ]
    const names = [
      'Blue'
    ]
    const markAllAsRead = (): void => window.alert('Callback on mark all as read')
    const messages: MessageCustom[] = [{
      messageType: CUSTOM_MESSAGE,
      details: {
        channel: 'channel-k63pjit0',
        from: {
          force: 'Red',
          forceColor: '#F00',
          iconURL: 'default_img/umpireDefault.png',
          roleName: 'CO',
          roleId: 'u345'
        },
        messageType: 'Chat',
        timestamp: '2020-10-13T08:52:04.394Z',
        turnNumber: 1
      },
      message: {
        content: 'message from Red'
      },
      _id: '2020-03-25T15:08:47.520Z',
      _rev: '1',
      hasBeenRead: false,
      isOpen: false
    }]

    const tree = renderer
      .create(<ChannelMessagesList messages={messages} names={names} onRead={undefined} onUnread={undefined} isUmpire={true} role={'Comms'} playerForceId={'Blue'} colors={colors} icons={icons} onMarkAllAsRead={markAllAsRead} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
