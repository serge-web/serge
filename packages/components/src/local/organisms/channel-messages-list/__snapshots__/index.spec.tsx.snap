// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelMessagesList component:  renders component correctly 1`] = `
<div>
  <div
    className="MuiBox-root MuiBox-root-1"
  >
    <div
      className="forces-in-channel"
    >
      <img
        alt=""
        className="force-indicator role-icon"
        src="./images/default_img/forceDefault.png"
        style={
          Object {
            "backgroundColor": "#F00",
            "borderColor": "#F00",
          }
        }
      />
      <button
        className="MuiButtonBase-root MuiButton-root makeStyles-root-2 MuiButton-text MuiButton-textSecondary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Mark all read
        </span>
      </button>
    </div>
  </div>
  <div
    className="MuiBox-root MuiBox-root-3 messages-list"
  >
    <div
      className="MuiBox-root MuiBox-root-4"
    >
      <div
        className="message-list-wrapper"
        style={
          Object {
            "borderColor": "#ffffff",
          }
        }
      >
        <div
          className="main"
          onClick={[Function]}
        >
          <div
            className=""
            onClick={[Function]}
          >
            <div
              className="message-title-wrap "
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                focusable="false"
                style={
                  Object {
                    "color": "#18ffff",
                  }
                }
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
              <div
                className="message-title"
              >
                lorem ipsum do lor sit amet
              </div>
              <div
                className="MuiBox-root MuiBox-root-5"
              >
                <span
                  className="info-body"
                >
                  11:11
                </span>
                <div
                  className="MuiChip-root makeStyles-root-6 makeStyles-root-9 MuiChip-sizeSmall makeStyles-sizeSmall-7"
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                >
                  <span
                    className="MuiChip-label makeStyles-label-8 makeStyles-label-10 MuiChip-labelSmall"
                  >
                    Game Control
                  </span>
                </div>
                <div
                  className="MuiChip-root makeStyles-root-6 makeStyles-root-11 MuiChip-sizeSmall makeStyles-sizeSmall-7"
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                >
                  <span
                    className="MuiChip-label makeStyles-label-8 makeStyles-label-12 MuiChip-labelSmall"
                  >
                    State of The World
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            className="wrap-detail "
          >
            <span
              className="detail"
            >
              <div
                className="MuiChip-root makeStyles-root-13 makeStyles-root-16 MuiChip-sizeSmall makeStyles-sizeSmall-14"
                onKeyDown={[Function]}
                onKeyUp={[Function]}
              >
                <span
                  className="MuiChip-label MuiChip-labelSmall makeStyles-labelSmall-15"
                >
                  Forces:
                </span>
              </div>
            </span>
            <p
              className="detail-rows"
            >
              <p
                className="detail-row"
              >
                <span
                  className="detail"
                >
                  <div
                    className="MuiChip-root makeStyles-root-13 makeStyles-root-17 MuiChip-sizeSmall makeStyles-sizeSmall-14"
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                  >
                    <span
                      className="MuiChip-label MuiChip-labelSmall makeStyles-labelSmall-15"
                    >
                      assets:
                    </span>
                  </div>
                </span>
                <p
                  className="detail-rows"
                >
                  <p
                    className="detail-row"
                  >
                    <span
                      className="detail"
                    >
                      <div
                        className="MuiChip-root makeStyles-root-13 makeStyles-root-18 MuiChip-sizeSmall makeStyles-sizeSmall-14"
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                      >
                        <span
                          className="MuiChip-label MuiChip-labelSmall makeStyles-labelSmall-15"
                        >
                          location:
                        </span>
                      </div>
                    </span>
                    <span
                      className="data"
                    >
                      loc
                    </span>
                    <span
                      className="detail"
                    >
                      <div
                        className="MuiChip-root makeStyles-root-13 makeStyles-root-19 MuiChip-sizeSmall makeStyles-sizeSmall-14"
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                      >
                        <span
                          className="MuiChip-label MuiChip-labelSmall makeStyles-labelSmall-15"
                        >
                          name:
                        </span>
                      </div>
                    </span>
                    <span
                      className="data"
                    >
                      name
                    </span>
                    <span
                      className="detail"
                    >
                      <div
                        className="MuiChip-root makeStyles-root-13 makeStyles-root-20 MuiChip-sizeSmall makeStyles-sizeSmall-14"
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                      >
                        <span
                          className="MuiChip-label MuiChip-labelSmall makeStyles-labelSmall-15"
                        >
                          visibleTo:
                        </span>
                      </div>
                    </span>
                    <p
                      className="detail-rows"
                    >
                      <p
                        className="detail-row"
                      >
                        <span
                          className="detail"
                        >
                          <div
                            className="MuiChip-root makeStyles-root-13 makeStyles-root-21 MuiChip-sizeSmall makeStyles-sizeSmall-14"
                            onKeyDown={[Function]}
                            onKeyUp={[Function]}
                          >
                            <span
                              className="MuiChip-label MuiChip-labelSmall makeStyles-labelSmall-15"
                            >
                              Force:
                            </span>
                          </div>
                        </span>
                        <span
                          className="data"
                        >
                          Blue
                        </span>
                      </p>
                    </p>
                  </p>
                  <p
                    className="detail-row"
                  >
                    <span
                      className="detail"
                    >
                      <div
                        className="MuiChip-root makeStyles-root-13 makeStyles-root-22 MuiChip-sizeSmall makeStyles-sizeSmall-14"
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                      >
                        <span
                          className="MuiChip-label MuiChip-labelSmall makeStyles-labelSmall-15"
                        >
                          location:
                        </span>
                      </div>
                    </span>
                    <span
                      className="data"
                    >
                      lo2
                    </span>
                    <span
                      className="detail"
                    >
                      <div
                        className="MuiChip-root makeStyles-root-13 makeStyles-root-23 MuiChip-sizeSmall makeStyles-sizeSmall-14"
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                      >
                        <span
                          className="MuiChip-label MuiChip-labelSmall makeStyles-labelSmall-15"
                        >
                          name:
                        </span>
                      </div>
                    </span>
                    <span
                      className="data"
                    >
                      name2
                    </span>
                    <span
                      className="detail"
                    >
                      <div
                        className="MuiChip-root makeStyles-root-13 makeStyles-root-24 MuiChip-sizeSmall makeStyles-sizeSmall-14"
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                      >
                        <span
                          className="MuiChip-label MuiChip-labelSmall makeStyles-labelSmall-15"
                        >
                          visibleTo:
                        </span>
                      </div>
                    </span>
                    <p
                      className="detail-rows"
                    >
                      <p
                        className="detail-row"
                      >
                        <span
                          className="detail"
                        >
                          <div
                            className="MuiChip-root makeStyles-root-13 makeStyles-root-25 MuiChip-sizeSmall makeStyles-sizeSmall-14"
                            onKeyDown={[Function]}
                            onKeyUp={[Function]}
                          >
                            <span
                              className="MuiChip-label MuiChip-labelSmall makeStyles-labelSmall-15"
                            >
                              Force:
                            </span>
                          </div>
                        </span>
                        <span
                          className="data"
                        >
                          Red
                        </span>
                      </p>
                    </p>
                  </p>
                </p>
                <span
                  className="detail"
                >
                  <div
                    className="MuiChip-root makeStyles-root-13 makeStyles-root-26 MuiChip-sizeSmall makeStyles-sizeSmall-14"
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                  >
                    <span
                      className="MuiChip-label MuiChip-labelSmall makeStyles-labelSmall-15"
                    >
                      force:
                    </span>
                  </div>
                </span>
                <span
                  className="data"
                >
                  Blue
                </span>
              </p>
              <p
                className="detail-row"
              >
                <span
                  className="detail"
                >
                  <div
                    className="MuiChip-root makeStyles-root-13 makeStyles-root-27 MuiChip-sizeSmall makeStyles-sizeSmall-14"
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                  >
                    <span
                      className="MuiChip-label MuiChip-labelSmall makeStyles-labelSmall-15"
                    >
                      assets:
                    </span>
                  </div>
                </span>
                <p
                  className="detail-rows"
                >
                  <p
                    className="detail-row"
                  >
                    <span
                      className="detail"
                    >
                      <div
                        className="MuiChip-root makeStyles-root-13 makeStyles-root-28 MuiChip-sizeSmall makeStyles-sizeSmall-14"
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                      >
                        <span
                          className="MuiChip-label MuiChip-labelSmall makeStyles-labelSmall-15"
                        >
                          location:
                        </span>
                      </div>
                    </span>
                    <span
                      className="data"
                    >
                      loc
                    </span>
                    <span
                      className="detail"
                    >
                      <div
                        className="MuiChip-root makeStyles-root-13 makeStyles-root-29 MuiChip-sizeSmall makeStyles-sizeSmall-14"
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                      >
                        <span
                          className="MuiChip-label MuiChip-labelSmall makeStyles-labelSmall-15"
                        >
                          name:
                        </span>
                      </div>
                    </span>
                    <span
                      className="data"
                    >
                      name3
                    </span>
                    <span
                      className="detail"
                    >
                      <div
                        className="MuiChip-root makeStyles-root-13 makeStyles-root-30 MuiChip-sizeSmall makeStyles-sizeSmall-14"
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                      >
                        <span
                          className="MuiChip-label MuiChip-labelSmall makeStyles-labelSmall-15"
                        >
                          visibleTo:
                        </span>
                      </div>
                    </span>
                    <p
                      className="detail-rows"
                    />
                  </p>
                </p>
                <span
                  className="detail"
                >
                  <div
                    className="MuiChip-root makeStyles-root-13 makeStyles-root-31 MuiChip-sizeSmall makeStyles-sizeSmall-14"
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                  >
                    <span
                      className="MuiChip-label MuiChip-labelSmall makeStyles-labelSmall-15"
                    >
                      force:
                    </span>
                  </div>
                </span>
                <span
                  className="data"
                >
                  Red
                </span>
              </p>
            </p>
            <div
              className="wrap-private"
            >
              <span
                className="detail"
              >
                <div
                  className="MuiChip-root makeStyles-root-13 makeStyles-root-32 MuiChip-sizeSmall makeStyles-sizeSmall-14"
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                >
                  <span
                    className="MuiChip-label MuiChip-labelSmall makeStyles-labelSmall-15"
                  >
                    <span>
                      <span
                        className="icon-private"
                      >
                        <svg
                          aria-hidden="true"
                          className="svg-inline--fa fa-user-secret fa-w-14 fa-1x "
                          data-icon="user-secret"
                          data-prefix="fas"
                          focusable="false"
                          role="img"
                          style={Object {}}
                          viewBox="0 0 448 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M383.9 308.3l23.9-62.6c4-10.5-3.7-21.7-15-21.7h-58.5c11-18.9 17.8-40.6 17.8-64v-.3c39.2-7.8 64-19.1 64-31.7 0-13.3-27.3-25.1-70.1-33-9.2-32.8-27-65.8-40.6-82.8-9.5-11.9-25.9-15.6-39.5-8.8l-27.6 13.8c-9 4.5-19.6 4.5-28.6 0L182.1 3.4c-13.6-6.8-30-3.1-39.5 8.8-13.5 17-31.4 50-40.6 82.8-42.7 7.9-70 19.7-70 33 0 12.6 24.8 23.9 64 31.7v.3c0 23.4 6.8 45.1 17.8 64H56.3c-11.5 0-19.2 11.7-14.7 22.3l25.8 60.2C27.3 329.8 0 372.7 0 422.4v44.8C0 491.9 20.1 512 44.8 512h358.4c24.7 0 44.8-20.1 44.8-44.8v-44.8c0-48.4-25.8-90.4-64.1-114.1zM176 480l-41.6-192 49.6 32 24 40-32 120zm96 0l-32-120 24-40 49.6-32L272 480zm41.7-298.5c-3.9 11.9-7 24.6-16.5 33.4-10.1 9.3-48 22.4-64-25-2.8-8.4-15.4-8.4-18.3 0-17 50.2-56 32.4-64 25-9.5-8.8-12.7-21.5-16.5-33.4-.8-2.5-6.3-5.7-6.3-5.8v-10.8c28.3 3.6 61 5.8 96 5.8s67.7-2.1 96-5.8v10.8c-.1.1-5.6 3.2-6.4 5.8z"
                            fill="currentColor"
                            style={Object {}}
                          />
                        </svg>
                      </span>
                      Private:
                    </span>
                  </span>
                </div>
              </span>
              <span
                className="private"
              >
                <div
                  className="paragraph"
                >
                  Private message
                </div>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
